kernel/src/engine/sync/mod.rs

```diff
@@ -60,12 +60,12 @@ impl Engine for SyncEngine {
 fn read_files<F, I>(
     files: &[FileMeta],
     schema: SchemaRef,
-    predicate: Option<Expression>,
+    predicate: Option<ExpressionRef>,
     mut try_create_from_file: F,
 ) -> DeltaResult<FileDataReadResultIterator>
 where
     I: Iterator<Item = DeltaResult<ArrowEngineData>> + Send + 'static,
-    F: FnMut(File, SchemaRef, ArrowSchemaRef, Option<&Expression>) -> DeltaResult<I>
```

Comment:
checking my understanding: in places where we previously took `&Expression` we are switching to Arc to allow for the expression 'sharing' we want? to have this expression share among many places via Arc? if we left as &Expression we would have to have an owned one somewhere which we are handing out references to?

Reply:
That is a (small) part of it. The main reason is that iterators tend to outlive the method invocation that creates them and that makes it really difficult to track lifetimes of a borrowed reference the iterator needs to consume. And it's outright impossible if there's any async code lurking underneath.